File: Sources/Sentry/SentryViewPhotographer.m
#import "SentryViewPhotographer.h"

#if SENTRY_HAS_UIKIT
NS_ASSUME_NONNULL_BEGIN

@implementation SentryViewPhotographer {
    NSMutableArray<Class> *_ignoreClasses;
    NSMutableArray<Class> *_redactClasses;
}

+ (SentryViewPhotographer *)shared
{
    static SentryViewPhotographer *_shared = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{ _shared = [[SentryViewPhotographer alloc] init]; });

    return _shared;
}

- (instancetype)init
{
    if (self = [super init]) {
#if TARGET_OS_IOS
        _ignoreClasses = @[ UISlider.class, UISwitch.class].mutableCopy;
#endif
        _redactClasses = @[ UILabel.class, UITextView.class, UITextField.class ].mutableCopy;

        NSArray<NSString *> *extraClasses = @[
            @"_TtCOCV7SwiftUI11DisplayList11ViewUpdater8Platform13CGDrawingView",
            @"_TtC7SwiftUIP33_A34643117F00277B93DEBAB70EC0697122_UIShapeHitTestingView",
            @"SwiftUI._UIGraphicsView", @"SwiftUI.ImageLayer"
        ];

        for (NSString *className in extraClasses) {
            Class viewClass = NSClassFromString(className);
            if (viewClass != nil) {
                [_redactClasses addObject:viewClass];
            }
        }
    }
    return self;
}

- (UIImage *)imageFromUIView:(UIView *)view
{
    UIGraphicsBeginImageContextWithOptions(view.bounds.size, YES, 0);
    CGContextRef currentContext = UIGraphicsGetCurrentContext();

    [view.layer renderInContext:currentContext];

    [self maskText:view context:currentContext];

    UIImage *screenshot = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();

    return screenshot;
}

- (void)maskText:(UIView *)view context:(CGContextRef)context
{
    [UIColor.blackColor setFill];
    CGPathRef maskPath = [self buildPathForView:view
                                         inPath:CGPathCreateMutable()
                                    visibleArea:view.frame];
    CGContextAddPath(context, maskPath);
    CGContextFillPath(context);
}

- (BOOL)shouldIgnoreView:(UIView *)view
{
    return [_ignoreClasses containsObject:view.class];
}

- (BOOL)shouldIgnore:(UIView *)view
{
    for (Class class in _ignoreClasses) {
        if ([view isKindOfClass:class]) {
            return true;
        }
    }
    return false;
}

- (BOOL)shouldRedact:(UIView *)view
{
    for (Class class in _redactClasses) {
        if ([view isKindOfClass:class]) {
            return true;
        }
    }

    return (
        [view isKindOfClass:UIImageView.class] && [self shouldRedactImageView:(UIImageView *)view]);
}

- (BOOL)shouldRedactImageView:(UIImageView *)imageView
{
    return imageView.image != nil &&
        [imageView.image.imageAsset valueForKey:@"_containingBundle"] == nil
        && (imageView.image.size.width > 10
            && imageView.image.size.height > 10); // This is to avoid redact gradient backgroud that
                                                  // are usually small lines repeating
}

- (CGMutablePathRef)buildPathForView:(UIView *)view
                              inPath:(CGMutablePathRef)path
                         visibleArea:(CGRect)area
{
    CGRect rectInWindow = [view convertRect:view.bounds toView:nil];

    if (!CGRectIntersectsRect(area, rectInWindow)) {
        return path;
    }

    if (view.hidden || view.alpha == 0) {
        return path;
    }

    BOOL ignore = [self shouldIgnore:view];
    if (!ignore && [self shouldRedact:view]) {
        CGPathAddRect(path, NULL, rectInWindow);
        return path;
    } else if ([self isOpaqueOrHasBackground:view]) {
        CGMutablePathRef newPath = [self excludeRect:rectInWindow fromPath:path];
        CGPathRelease(path);
        path = newPath;
    }

    if (!ignore) {
        for (UIView *subview in view.subviews) {
            path = [self buildPathForView:subview inPath:path visibleArea:area];
        }
    }

    return path;
}

- (CGMutablePathRef)excludeRect:(CGRect)rectangle fromPath:(CGMutablePathRef)path
{
    if (@available(iOS 16.0,tvOS 16.0, *)) {
        CGPathRef exclude = CGPathCreateWithRect(rectangle, nil);
        CGPathRef newPath = CGPathCreateCopyBySubtractingPath(path, exclude, YES);
        return CGPathCreateMutableCopy(newPath);
    }
    return path;
}

- (BOOL)isOpaqueOrHasBackground:(UIView *)view
{
    return view.isOpaque
        || (view.backgroundColor != nil && CGColorGetAlpha(view.backgroundColor.CGColor) > 0.9);
}

@end

NS_ASSUME_NONNULL_END
#endif // SENTRY_HAS_UIKIT

File: Sources/Swift/Integrations/SessionReplay/SentryViewPhotographer.swift
import Foundation

#if canImport(UIKit)
import CoreGraphics
import UIKit

@objcMembers
class SentryViewPhotographer: NSObject {
    
    private var ignoreClasses: [AnyClass] = []
    private var redactClasses: [AnyClass] = []
    
    static let shared = SentryViewPhotographer()
    
    override init() {
#if os(iOS)
        ignoreClasses = [  UISlider.self, UISwitch.self ]
#endif
        redactClasses = [ UILabel.self, UITextView.self, UITextField.self ] + [
            "_TtCOCV7SwiftUI11DisplayList11ViewUpdater8Platform13CGDrawingView",
            "_TtC7SwiftUIP33_A34643117F00277B93DEBAB70EC0697122_UIShapeHitTestingView",
            "SwiftUI._UIGraphicsView", "SwiftUI.ImageLayer"
        ].compactMap { NSClassFromString($0) }
    }
    
    func image(view: UIView) -> UIImage? {
        UIGraphicsBeginImageContextWithOptions(view.bounds.size, true, 0)
        if let currentContext = UIGraphicsGetCurrentContext() {
            view.layer.render(in: currentContext)
            self.mask(view: view, context: currentContext)
            if let screenshot = UIGraphicsGetImageFromCurrentImageContext() {
                UIGraphicsEndImageContext()
                return screenshot
            }
        }
        UIGraphicsEndImageContext()
        return nil
    }
    
    private func mask(view: UIView, context: CGContext) {
        UIColor.black.setFill()
        let maskPath = self.buildPath(view: view, path: CGMutablePath(), area: view.frame)
        context.addPath(maskPath)
        context.fillPath()
    }
    
    private func shouldIgnore(view: UIView) -> Bool {
        ignoreClasses.contains { view.isKind(of: $0) }
    }
    
    private func shouldRedact(view: UIView) -> Bool {
        if let imageView = view as? UIImageView {
            return shouldRedact(imageView: imageView)
        }
        
        return redactClasses.contains { view.isKind(of: $0) }
    }
    
    private func shouldRedact(imageView: UIImageView) -> Bool {
        // Checking the size is to avoid redact gradient backgroud that
        // are usually small lines repeating
        guard let image = imageView.image, image.size.width > 10 && image.size.height > 10  else { return false }
        return image.imageAsset?.value(forKey: "_containingBundle") != nil
    }
    
    private func buildPath(view: UIView, path: CGMutablePath, area: CGRect) -> CGMutablePath {
        let rectInWindow = view.convert(view.bounds, to: nil)

        if !area.intersects(rectInWindow) || view.isHidden || view.alpha == 0 {
            return path
        }
        
        var result = path

        let ignore = shouldIgnore(view: view)
        
        if !ignore && shouldRedact(view: view) {
            result.addRect(rectInWindow)
            return result
        } else if isOpaqueOrHasBackground(view) {
            let newPath = excludeRect(rectInWindow, fromPath: result)
            result = newPath
        }

        if !ignore {
            for subview in view.subviews {
                result = buildPath(view: subview, path: path, area: area)
            }
        }

        return result
    }
    
    private func excludeRect(_ rectangle: CGRect, fromPath path: CGMutablePath) -> CGMutablePath {
        if #available(iOS 16.0, tvOS 16.0, *) {
            let exclude = CGPath(rect: rectangle, transform: nil)
            let newPath = path.subtracting(exclude, using: .evenOdd)
            return newPath.mutableCopy() ?? path
        }
        return path
    }
    
    private func isOpaqueOrHasBackground(_ view: UIView) -> Bool {
        return view.isOpaque || (view.backgroundColor != nil && (view.backgroundColor?.cgColor.alpha ?? 0) > 0.9)
    }
}

#endif // canImport(UIKit)
